### PR TITLE
MFCC features & featurizer refactoring

### PR DESCRIPTION
**Pull request by redoctopus**
Moved audio preprocessing featurizers into their own file, rearranged the class structure, and replaced some code with torchaudio.transforms, as discussed previously.

This refactor splits AudioPreprocessing into a base class + three NMs, one per transform: audio -> [spectrogram, mel spec, MFCC].

I'm holding off on more granularity (i.e. intermediate transforms) for now because we only support loading audio and kaldi MFCCs (which don't need this preprocessing), and as far as I know don't have plans to support loading intermediate features. We can revisit discussion about structuring the preprocessors if we decide to in the future.

### CODE REVIEW

File: examples/asr/configs/quartznet_an4.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 model: "QuartzNet-AN4"
-sample_rate: &sr 16000
```
Comment (blisc) [inline examples/asr/configs/quartznet_an4.yaml]: Why was sample_rate removed?

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+from .parts.spectr_augment import SpecAugment, SpecCutout
+
+
+class AudioPreprocessor(NonTrainableNM):
+    """
+    A base class for Neural Modules that performs audio preprocessing,
+    transforming the wav files to features.
+    """
+    @staticmethod
+    def create_ports():
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: This class shouldn't have ports

Comment (blisc) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Each child class should be returning ports with different tags. As part of nemo_tts, I added MelSpectrogramSignalTag. Please add one for MFCC too.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+        normalized (bool): Whether to normalize by magnitude after stft
+        pad (int): Two sided padding
+    """
+    def __init__(
+            self, *,
+            sample_rate=16000,
+            window_size=0.02,
+            window_stride=0.01,
+            n_fft=None,
+            window="hamming",
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: why hamming? In general, why do we have different default windows for all three processors

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+            self, *,
+            sample_rate=16000,
+            window_size=None,
+            window_stride=None,
+            window=None,
+            n_mfcc=64,
+            dct_type=2,
+            norm='ortho',
+            log=True,
+            mel_kwargs=None,
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Please don't pass dictionaries to functions. If we cannot express the functionality as a parameter, I would prefer we just don't have that functionality rather than allow it to be passed as a dictionary.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+            with amp.disable_casts():
+                processed_signal = self.get_features(
+                        input_signal.to(torch.float), length)
+        else:
+            processed_signal = self.get_features(input_signal, length)
+
+        processed_length = self.get_seq_len(length)
+        return processed_signal, processed_length
+
+    def get_features(self, input_signal, length):
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Please make this abstract method instead.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+        # Called by forward(). Subclasses should implement this.
+        raise NotImplementedError
+
+    def get_seq_len(self, length):
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Now this is a good class to not make abstract because all three children essentially are call the same get_seq_len function.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+    """
+    def __init__(
+            self, *,
+            sample_rate=16000,
+            window_size=0.02,
+            window_stride=0.01,
+            n_fft=None,
+            window="hamming",
+            normalized=True,
+            pad=8,
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Are you sure we want two-side padding here? This is audio padding, not spectrogram padding.
Also please be aware that torch.stft() has different defaults for padding than torchaudio.transforms.Spectrogram().
I personally would prefer the torch.stft() defaults over torchaudio.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        self.win_length = int(sample_rate * window_size)
+        self.hop_length = int(sample_rate * window_stride)
+        self.n_fft = n_fft or 2 ** math.ceil(math.log2(self.win_length))
+
+        window_fn = self.torch_windows.get(window, None)
+        if window_fn is None:
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: A None window type should be allowed. AKA your window is just 1.
However, an unknown window type should throw this error. This is a stuble difference depending on what you intend self.torch_windows["none"] to mean.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+        # Override mel_kwargs if window_size or window_stride are set
+        if window_size is not None:
+            mel_kwargs['win_length'] = int(sample_rate * window_size)
+        if window_stride is not None:
+            mel_kwargs['hop_length'] = int(sample_rate * window_stride)
+
+        # Set window_fn if window arg is given
+        if window:
+            window_fn = self.torch_windows.get(window, None)
+            if window_fn is None:
```
Comment (blisc) [inline collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Same comment as above.
A None window type should be allowed. AKA your window is just 1.
However, an unknown window type should throw this error. This is a stuble difference depending on what you intend self.torch_windows["none"] to mean.

**Comment by blisc** [comment]
This will break nemo_tts, can you please merge with the latest master and fix conflicts so unittests can be run?

**Comment by redoctopus** [comment]
Fixing conflicts before addressing the other comments--do you mind taking a look at my changes to nemo_tts to make sure they're fine? Tests passed but would like to double-check.

File: examples/asr/configs/quartznet_an4.yaml (hunk 1 above)
Comment (redoctopus) [reply examples/asr/configs/quartznet_an4.yaml]: Previously, passing this file as a config to `quartznet.py` would cause an error since this top-level sample_rate was passed in as well as the ones in the data layer params (i.e. duplicate sample_rate params passed in).
I just removed the two duplicate sample_rate params causing the issue below.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 3 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Hamming was the default for the spectrogram featurizer before I moved it over, and I'm not sure of the original reasoning. Will change this to 'hann' to match mel spec featurizer.
If we're flattening mel_kwargs for MFCC I'll change that as well.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 4 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: torchaudio uses a dictionary argument here.
So do we prefer to "flatten" it in the constructor and then create the dict within this class to pass in? Or not support it? I'd assume the former...

File: examples/asr/configs/quartznet_an4.yaml (hunk 1 above)
Comment (blisc) [reply examples/asr/configs/quartznet_an4.yaml]: Ok

**Comment by blisc** [comment]
I think tts_infer.py is still broken, but I can fix it after we merge this.
I think the other tts scripts should be fine, I'll do more testing after this.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 8 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: It doesn't seem like `torchaudio.transforms.Spectrogram` allows us to pass in `window_fn=None`. Should we just define our own lambda for a window of 1s, in this case? Or use the torchaudio default?

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 2 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 5 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 4 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done--expanded args into the AudioToMFCCPreprocessor.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 8 above)
Comment (blisc) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: torch.stft()'s default is None. They do the same, I don't know why they changed so much.
Would mapping "none" to torch.ones work?

Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done--mapped both "none" and "null" to torch.ones (some configs use null)

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 9 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done (see above)

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 3 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 8 above)
Comment (blisc) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Wait a second, a null value in a yaml file converts to None. This value is different from a string "null" and different from a string "none". Let's be careful with the self.torch_windows dictionary. I would advise to remove both strings. I would allow "ones" to map to torch.ones.

We could add None: torch.ones in the self.torch_windows dict, but I am not sure about this.

Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Added None to the self.torch_windows dict and removed 'null' and 'none' as per in-person conversation

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 7 above)
Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Removed pad option here according to discussion.
Turns out torchaudio calls torch.stft, and hard-codes the center arg to True anyway.

File: collections/nemo_asr/nemo_asr/audio_preprocessing.py (hunk 6 above)
Comment (blisc) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Let's make define this function for the base class as `return torch.ceil(seq_len / self.hop_length).to(dtype=torch.long)` and remove this function from all children.
Let's make the base class init require either n_window_stride or hop_length and set self.hop_length inside the init.

Comment (redoctopus) [reply collections/nemo_asr/nemo_asr/audio_preprocessing.py]: Done.
